packages/gradle/src/plugin/nodes.ts

```diff
@@ -0,0 +1,164 @@
+  writeJsonFile,
+} from '@nx/devkit';
+import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
+import { existsSync } from 'node:fs';
+import { dirname, join } from 'node:path';
+import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';
+
+import { getGradleBinaryPath } from '../utils/exec-gradle';
+import { getGradleReport } from '../utils/get-gradle-report';
+
+const nonCacheableGradleTaskTypes = new Set(['Application']);
+const dependsOnMap = {
+  build: ['^build', 'classes'],
+  test: ['classes'],
+  classes: ['^classes'],
+};
+
+interface GradleTask {
+  type: string;
+  name: string;
+}
+
+export interface GradlePluginOptions {
+  testTargetName?: string;
+  classesTargetName?: string;
+  buildTargetName?: string;
+  [taskTargetName: string]: string | undefined;
+}
+
+const cachePath = join(projectGraphCacheDirectory, 'gradle.hash');
+const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};
+
+export const calculatedTargets: Record<
+  string,
+  Record<string, TargetConfiguration>
+> = {};
+
+function readTargetsCache(): Record<
+  string,
+  Record<string, TargetConfiguration>
+> {
+  return readJsonFile(cachePath);
+}
+
+export function writeTargetsToCache(
+  targets: Record<string, Record<string, TargetConfiguration>>
+) {
+  writeJsonFile(cachePath, targets);
+}
+
+export const createNodes: CreateNodes<GradlePluginOptions> = [
+  '{**/build.gradle.kts,**/build.gradle}',
+  (
+    gradleFilePath,
+    options: GradlePluginOptions | undefined,
+    context: CreateNodesContext
+  ) => {
+    const projectRoot = dirname(gradleFilePath);
+
+    const hash = calculateHashForCreateNodes(
+      projectRoot,
+      options ?? {},
+      context
+    );
+
+    try {
+      const {
+        tasksMap,
+        gradleProjectToTasksTypeMap,
+        gradleFileToOutputDirsMap,
+        gradleFileToGradleProjectMap,
+        gradleProjectToProjectName,
+      } = getGradleReport();
+
+      const gradleProject = gradleFileToGradleProjectMap.get(
+        gradleFilePath
+      ) as string;
+      const projectName = gradleProjectToProjectName.get(gradleProject);
+      if (!projectName) {
+        return;
+      }
+
+      const availableTaskNames = tasksMap.get(gradleFilePath) as string[];
+      const tasksTypeMap = gradleProjectToTasksTypeMap.get(
+        gradleProject
+      ) as Map<string, string>;
+      const tasks: GradleTask[] = availableTaskNames.map((taskName) => {
+        return {
+          type: tasksTypeMap.get(taskName) ?? 'Unknown',
+          name: taskName,
+        };
+      });
+
+      const outputDirs = gradleFileToOutputDirsMap.get(gradleFilePath) as Map<
+        string,
+        string
+      >;
+
+      const targets =
+        targetsCache[hash] ??
+        createGradleTargets(tasks, projectRoot, options, context, outputDirs);
+      calculatedTargets[hash] = targets;
+
+      return {
+        projects: {
+          [projectName]: {
```

Comment:
I think this should be` [projectRoot]: {....}`?

Reply:
changed